File: python/keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = set()
        keys_stack = [0]
        while keys_stack:
            key = keys_stack.pop()
            if key in visited:
                continue
            visited.add(key)
            for new_key in rooms[key]:
                keys_stack.append(new_key)
        return len(visited) == len(rooms)

File: go/keys-and-rooms.go
func canVisitAllRooms(rooms [][]int) bool {
	keysStack := make([]int, 0)
	keysStack = append(keysStack, 0)
	visited := make([]bool, len(rooms))
	visited_count := 0
	for 0 < len(keysStack) {
		peak := len(keysStack) - 1
		key := keysStack[peak]
		keysStack = keysStack[:peak]
		if visited[key] {
			continue
		} else {
			visited_count += 1
			visited[key] = true
		}
		for i := 0; i < len(rooms[key]); i++ {
			keysStack = append(keysStack, rooms[key][i])
		}
	}
	return visited_count == len(rooms)
}
